### PR TITLE
start up without chainId (or network)

### PR DESCRIPTION
**Pull request by turbocrime**
this allows prax to init without a known chain id, by querying chain id from remote RPC, or using chain id from local storage.

as a result prax is more robust and can function while remote RPC is unavailable.

ui components such as sync progress also now function properly when the RPC is unavailable.

### CODE REVIEW

**Comment by turbocrime** [comment]
#971

File: packages/ui/components/ui/block-sync-status/condensed.tsx
Comment (turbocrime) [inline packages/ui/components/ui/block-sync-status/condensed.tsx:1]: block sync status components probably need to be using BigInt but that's out of scope for this pr

File: packages/ui/components/ui/block-sync-status/large.tsx
Comment (turbocrime) [inline packages/ui/components/ui/block-sync-status/large.tsx:1]: seems like this is unused

File: packages/storage/src/chrome/types.ts [hunk 1]
```diff
@@ -15,9 +17,10 @@ export interface OriginRecord {
...
-  grpcEndpoint?: string;
+  grpcEndpoint: string | undefined;
   frontendUrl: string;
-  passwordKeyPrint?: KeyPrintJson;
-  fullSyncHeight?: number;
+  passwordKeyPrint: KeyPrintJson | undefined;
+  fullSyncHeight: number | undefined;
   knownSites: OriginRecord[];
+  params: Jsonified<AppParameters> | undefined;
 }
```
Comment (turbocrime) [inline packages/storage/src/chrome/types.ts:26]: in order to successfully return undefined from the storage getter, undefined had to be explicitly added to the state type.

File: packages/query/src/queriers/shielded-pool.ts [hunk 2]
```diff
@@ -14,8 +14,9 @@ export class ShieldedPoolQuerier implements ShieldedPoolQuerierInterface {
...
 
-  async assetMetadataById(assetId: AssetId): Promise<Metadata | undefined> {
-    const res = await this.client.assetMetadataById({ assetId });
-    return res.denomMetadata;
-  }
+  assetMetadataById = (assetId: AssetId): Promise<Metadata | undefined> =>
+    this.client.assetMetadataById({ assetId }).then(
+      ({ denomMetadata }) => denomMetadata,
+      () => undefined,
+    );
```
[the remark this answers is not in the document]
Comment (grod220) [reply packages/query/src/queriers/shielded-pool.ts:21]: I'd say the same for `latestBlockHeight()` below too

File: packages/services-context/src/index.ts [hunk 3]
```diff
@@ -88,12 +90,32 @@ export class Services implements ServicesInterface {
...
+    const queriedParams = await this.querier.app.appParams().catch((e: unknown) => void e);
+
+    // prefer fetched params
+    const chainId = queriedParams?.chainId ?? storedParams?.chainId;
+    const epochDuration =
+      queriedParams?.sctParams?.epochDuration ?? storedParams?.sctParams?.epochDuration;
+
+    // store chainId if it was missing
+    if (!storedParams?.chainId && queriedParams?.chainId)
+      await localExtStorage.set('params', queriedParams.toJson() as Jsonified<AppParameters>);
```
Comment (grod220) [inline packages/services-context/src/index.ts:108]: Brackets here please

File: packages/services-context/src/index.ts [hunk 4]
```diff
@@ -88,12 +90,32 @@ export class Services implements ServicesInterface {
...
+    const storedParams = await localExtStorage
+      .get('params')
+      .then(j => j && AppParameters.fromJson(j));
+    // try to fetch params from network
+    const queriedParams = await this.querier.app.appParams().catch((e: unknown) => void e);
+
+    // prefer fetched params
+    const chainId = queriedParams?.chainId ?? storedParams?.chainId;
+    const epochDuration =
+      queriedParams?.sctParams?.epochDuration ?? storedParams?.sctParams?.epochDuration;
```
Comment (grod220) [inline packages/services-context/src/index.ts:104]: Can we move the asserts closer to where the variables are set? Something like:

```ts
    const chainId = queriedParams?.chainId ?? storedParams?.chainId;
    if (!chainId) throw new Error('Cannot initialize without chainId');
    // fail if chainId is mismatched
    if (storedParams?.chainId && storedParams.chainId !== chainId)
      throw new Error(
        'Stored chainId does not match the remote chainId. Your local state may\
        be invalid, and you cannot use this remote endpoint without clearing\
        your cached chain.',
      );

    const epochDuration =
      queriedParams?.sctParams?.epochDuration ?? storedParams?.sctParams?.epochDuration;
    if (!epochDuration) throw new Error('Cannot initialize without epoch duration');

    // store chainId if it was missing
    if (!storedParams?.chainId && queriedParams?.chainId) {
      await localExtStorage.set('params', queriedParams.toJson() as Jsonified<AppParameters>);
    }
```

File: packages/services/src/view-service/status-stream.ts [hunk 5]
```diff
@@ -4,18 +4,14 @@ import { servicesCtx } from '../ctx/prax';
...
-  }
+  // This should stream forever unless cancelled.
+  let remoteBlockHeight: bigint | undefined;
+  for await (const { value: syncHeight } of indexedDb.subscribe('FULL_SYNC_HEIGHT'))
+    if ((remoteBlockHeight ??= await services.querier.tendermint.latestBlockHeight()))
+      yield {
+        latestKnownBlockHeight: syncHeight <= remoteBlockHeight ? remoteBlockHeight : syncHeight,
+        partialSyncHeight: syncHeight,
+        fullSyncHeight: syncHeight,
+      };
```
Comment (grod220) [inline packages/services/src/view-service/status-stream.ts:16]: The lack of brackets make this hard to understand. Don't think we should ever do bracket-less loops.

File: packages/services/src/view-service/status-stream.ts [hunk 6]
```diff
@@ -4,18 +4,14 @@ import { servicesCtx } from '../ctx/prax';
...
-    yield {
-      latestKnownBlockHeight: syncHeight >= latestBlockHeight ? syncHeight : latestBlockHeight,
-      partialSyncHeight: syncHeight,
-      fullSyncHeight: syncHeight,
-    };
-  }
+  // This should stream forever unless cancelled.
+  let remoteBlockHeight: bigint | undefined;
+  for await (const { value: syncHeight } of indexedDb.subscribe('FULL_SYNC_HEIGHT'))
+    if ((remoteBlockHeight ??= await services.querier.tendermint.latestBlockHeight()))
```
Comment (grod220) [inline packages/services/src/view-service/status-stream.ts:11]: Think this is a bit more understandable if we can pull this out, aka:

```ts
  for await (const { value: syncHeight } of indexedDb.subscribe('FULL_SYNC_HEIGHT')) {
    remoteBlockHeight ??= await services.querier.tendermint.latestBlockHeight();
    if (remoteBlockHeight) {
      yield {
```

File: packages/services-context/src/index.ts (hunk 3 above)
Comment (turbocrime) [reply packages/services-context/src/index.ts:108]: i'm going to make a PR fixing the eslint config to reenable react-hooks, and i'll add a linter rule for braces at that point

File: packages/services-context/src/index.ts (hunk 4 above)
Comment (turbocrime) [reply packages/services-context/src/index.ts:104]: idb can start up without epochDuration, so they are deliberately separated. i'll pull out the params request into a method to simplify it conceptually

File: packages/services-context/src/index.ts
Comment (turbocrime) [inline packages/services-context/src/index.ts:1]: full config is now guaranteed at init, so it doesn't need the promise logic anymore

**Pull request merged**